easybuild/tools/containers.py

```diff
@@ -0,0 +1,304 @@
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Support for generating container recipes and creating container images
+
+:author: Shahzeb Siddiqui (Pfizer)
+:author: Kenneth Hoste (HPC-UGent)
+"""
+import os
+from distutils.version import LooseVersion
+from vsc.utils import fancylogger
+
+from easybuild.tools.build_log import EasyBuildError, print_msg
+from easybuild.tools.config import CONT_IMAGE_FORMAT_EXT3, CONT_IMAGE_FORMAT_SANDBOX, CONT_IMAGE_FORMAT_SQUASHFS
+from easybuild.tools.config import build_option, container_path
+from easybuild.tools.filetools import which, write_file
+from easybuild.tools.module_naming_scheme.utilities import det_full_ec_version
+from easybuild.tools.run import run_cmd
+
+
+CONT_TYPE_DOCKER = 'docker'
+CONT_TYPE_SINGULARITY = 'singularity'
+
+DOCKER = 'docker'
+LOCALIMAGE = 'localimage'
+SHUB = 'shub'
+SINGULARITY_BOOTSTRAP_TYPES = [DOCKER, LOCALIMAGE, SHUB]
+
+
+SINGULARITY_TEMPLATE = """
+Bootstrap: %(bootstrap)s
+From: %(from)s
+
+%%post
+%(install_os_deps)s
+
+# upgrade easybuild package automatically to latest version
+pip install -U easybuild
+
+# change to 'easybuild' user
+su - easybuild
+
+eb %(easyconfigs)s --robot --installpath=/app/ --prefix=/scratch --tmpdir=/scratch/tmp
+
+# exit from 'easybuild' user
+exit
+
+# cleanup
+rm -rf /scratch/tmp/* /scratch/build /scratch/sources /scratch/ebfiles_repo
+
+%%runscript
+eval "$@"
+
+%%environment
+source /etc/profile
+module use /app/modules/all
+module load %(mod_names)s
+
+%%labels
+
+"""
+
+
+_log = fancylogger.getLogger('tools.containers')  # pylint: disable=C0103
+
+
+def parse_container_base(base):
+    """Parse value passed to --container-base option."""
+    if base:
+        base_specs = base.split(':')
+        if len(base_specs) > 3 or len(base_specs) <= 1:
+            error_msg = '\n'.join([
+                "Invalid format for --container-base, must be one of the following:",
+                '',
+                "--container-base localimage:/path/to/image",
+                "--container-base shub:<image>:<tag>",
+                "--container-base docker:<image>:<tag>",
+            ])
+            raise EasyBuildError(error_msg)
+    else:
+        raise EasyBuildError("--container-base must be specified")
```

Comment:
@boegel So a first time user trying the eb-singularity stuff would have no clue what base container to use.  

```
[siddis14@amrndhl1157 easybuild-framework]$ eb M4-1.4.17.eb -C --experimental
== temporary log file in case of crash /tmp/eb-LxMMKc/easybuild-LHIN2Y.log
== Singularity tool found at /usr/local/bin/singularity
== Singularity version '2.4' is 2.4 or higher ... OK
ERROR: --container-base must be specified
```

Either add ``--container-base shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611`` and ``--container-base shub:shahzebsiddiqui/eb-singularity:centos-7.4.1708`` in the Error message or in the help section. Another option is if user don't specify ``--container-base`` easybuild can try to use one of the base containers either centos 7.3 or 7.4. In the future, I would like to move the base container into easybuilders repo instead of mine. But for now users have to use my container image since that is the only image available in SHUB.

Reply:
How about changing this to `ERROR: --container-base must be specified (see 'eb --help')`, or pointing to the documentation (URL to be figured out)?

since:

```
$ eb --help | grep container-base
    --container-base=BASE
                        --container-base localimage:/path/to/image.img,
                        --container-base shub:<image>:<tag>, --container-base
```

Reply:
I will document bootstrap easybuild containers for shub in the easybuild documentation. But users don't want to refer to documentation everytime to remember the SHUB easybuild base container since they will be using it.

It would be nice either to add the message in either eb --help or in the error message something like 

``` 
ERROR: --container-base must be specified. See below for examples

eb --container-base localimage:$HOME/.local/easybuild/containers/M4-1.4.17.simg
eb --container-base shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611
eb --container-base shub:shahzebsiddiqui/eb-singularity:centos-7.4.1708
eb --container-base docker:centos:7.3.1611
```

I'll let you decide on the messaging.

Reply:
`eb --help` already includes useful suggestions imho:

```
eb --help | grep container-base -A 1
    --container-base=BASE
                        Base for container image. Examples (for Singularity):
--
                        --container-base localimage:/path/to/image.img,
                        --container-base shub:<image>:<tag>, --container-base
                        docker:<image>:<tag>  (type <type 'str'>)
```